aiosupervisor/client.py

```diff
@@ -0,0 +1,203 @@
+        data: Any = None,
+    ) -> Response:
+        """Handle a request to Supervisor."""
+        url = URL(self.api_host).joinpath(uri)
+
+        match response_type:
+            case ResponseType.TEXT:
+                accept = "text/plain, */*"
+            case _:
+                accept = "application/json, text/plain, */*"
+
+        headers = {
+            "User-Agent": f"AioSupervisor/{VERSION}",
+            "Accept": accept,
+            "Authorization": f"Bearer {self.token}",
+        }
+
+        if self.session is None:
+            self.session = ClientSession()
+            self._close_session = True
+
+        try:
+            async with self.session.request(
+                method,
+                url,
+                timeout=self.timeout,
+                headers=headers,
+                params=params,
+                json=json,
+                data=data,
+            ) as response:
+                if response.status >= 400:
+                    exc_type: type[SupervisorError] = SupervisorError
+                    match response.status:
+                        case 400:
+                            exc_type = SupervisorBadRequestError
+                        case 401:
+                            exc_type = SupervisorAuthenticationError
+                        case 403:
+                            exc_type = SupervisorForbiddenError
+                        case 404:
+                            exc_type = SupervisorNotFoundError
+                        case 503:
+                            exc_type = SupervisorServiceUnavailableError
+
+                    if is_json(response):
+                        result = Response.from_json(await response.text())
+                        raise exc_type(result.message, result.job_id)
+                    raise exc_type()
+
+                match response_type:
+                    case ResponseType.JSON:
+                        is_json(response, raise_on_fail=True)
+                        return Response.from_json(await response.text())
+                    case ResponseType.TEXT:
+                        return Response(ResultType.OK, await response.text())
+                    case _:
+                        return Response(ResultType.OK)
+
+        except (UnicodeDecodeError, ClientResponseError) as err:
+            raise SupervisorResponseError(
+                "Unusable response received from Supervisor, check logs"
+            ) from err
+        except TimeoutError as err:
+            raise SupervisorTimeoutError("Timeout connecting to Supervisor") from err
+        except ClientError as err:
+            raise SupervisorConnectionError(
+                "Error occurred connecting to supervisor"
+            ) from err
+
+    def get(
+        self,
+        uri: str,
+        *,
+        params: dict[str, str] | None = None,
+        response_type: ResponseType = ResponseType.JSON,
+    ) -> Awaitable[Response]:
```

Comment:
Don't return awaitables. Make the methods coroutine functions and await `self._request`. We don't need to avoid coroutine chaining.

Reply:
Ah ok. Supervisor did that a lot, I assumed it was for performance reasons. If its not a big deal I'll change it here and maybe change that behavior in supervisor over time. Thanks!

Reply:
It's yielding to the event loop that is expensive. Just awaiting a coroutine function isn't expensive. I think this was a misunderstanding from some people in the team in the early days.